File: common/src/main/java/kg/delletenebre/yamus/api/responses/Playlist.kt
package kg.delletenebre.yamus.api.responses

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass


@JsonClass(generateAdapter = true)
data class Playlist(
    @Json(name = "available") val available: Boolean = true,
    @Json(name = "durationMs") val durationMs: Long = 0,
    @Json(name = "kind") val kind: Long = 0,
    @Json(name = "modified") val modified: String = "",
    @Json(name = "ogImage") val ogImage: String = "",
    @Json(name = "revision") val revision: Int = 0,
    @Json(name = "title") val title: String = "",
    @Json(name = "trackCount") val trackCount: Int = 0,
    @Json(name = "uid") val uid: Long = 0
)

File: lib/api/models/playlist.dart
import 'package:yamus/api/models.dart';

class Playlist {
  const Playlist({
    this.uid = '',
    this.kind = '',
    this.title = '',
    this.available = false,
    this.durationMs = 0,
    this.modified = '',
    this.coverUri = '',
    this.revision = 0,
    this.trackCount = 0,
    this.tracks = const [],
    this.trackIds = const [],
  });

  final String uid;
  final String kind;
  final String title;
  final bool available;
  final int durationMs;
  final String modified;
  final String coverUri;
  final int revision;
  final int trackCount;
  final List<Track> tracks;
  final List<TrackId> trackIds;

  factory Playlist.fromJson(Map<String, dynamic> json) {
    List<Track> tracks = [];
    List<TrackId> trackIds = [];
    if (json.containsKey('tracks')) {
      json['tracks'].forEach((item) {
        if (item.containsKey('track')) {
          tracks.add(Track.fromJson(item['track']));
        } else {
          trackIds.add(TrackId.fromJson(item));
        }
      });
    }

    return Playlist(
      uid: json['uid'].toString(),
      kind: json['kind'].toString(),
      title: json['title'],
      available: json['available'],
      durationMs: json['durationMs'],
      modified: json['modified'],
      coverUri: json['cover']['uri'],
      revision: json['revision'],
      trackCount: json['trackCount'],
      tracks: tracks,
      trackIds: trackIds,
    );
  }
}